### ISSUE TITLE
[libodb] Error building package on x64-linux

### ISSUE THREAD

**Issue by mcraveiro** [Microsoft/vcpkg#4428]
Hi vcpkg developers,

Thanks for a fantastic tool! I've managed to get up and running with boost and all of my other dependencies, so its all going great. However, when trying to add libodb I've hit a snag; it seems its not setup to compile on Linux. The problem, I think, comes from this line in ```ports/libodb/CMakeLists.txt```:

```
list(FILTER libodb_src EXCLUDE REGEX /posix/.*)
```

This removes all of the unix related code in libodb. When I try to compile it fails with an attempt to include ```windows.h```. I did a quick experiment and changed it to:

```
list(FILTER libodb_src EXCLUDE REGEX /win32/.*)
```

And the compilation worked fine. I am trying to investigate the right way to do this (presumably switch on some OS variable) but in the mean time, if someone is aware of the proper CMake/vcpkg way of doing this, please let me know.

Many thanks

Marco

**Comment by mcraveiro** [comment]
Right, I've knocked up a quick hack patch [1] that addresses the main issue - I can now compile on Linux. However, I think this raises a deeper question: from what I can tell, it seems that the vcpkg port uses a "hand-rolled" CMakeLists.txt [2] which is not part of the original ODB project [3]. Now, I may be getting confused, but if I'm not this does not seem to be the best idea as we now have a build of ODB which is not really being done as per the ODB project. So I guess I have a few questions:

1. Is it not possible to use the original automake files instead of creating a vcpkg specific CMakeList? Are there problems with using automake etc.
2. Is it worth submitting my changes as a PR? Its not ideal but it does build on Linux, so its better than the current situation, perhaps?

I'll continue to investigate, but if you have any ideas let me know.

Cheers

[1] https://github.com/MASD-Project/vcpkg/commit/6e6b412fcf1ba0aad16dd89a37910b4e4a82d90e
[2] https://github.com/MASD-Project/vcpkg/blob/master/ports/libodb/CMakeLists.txt
[3] https://git.codesynthesis.com/cgit/odb/libodb/tree/?h=v2.5.0-b.9

**Comment by mcraveiro** [comment]
OK, I've checked a few of the libraries [1], [2] and this indeed seems to be the convention. I'll try to tidy things up to match the examples and then submit a PR.

[1] https://github.com/MASD-Project/vcpkg/blob/master/ports/libxml2/CMakeLists.txt
[2] https://github.com/MASD-Project/vcpkg/blob/master/ports/libiconv/CMakeLists.txt

**Comment by ras0219-msft** [comment]
>Is it not possible to use the original automake files instead of creating a vcpkg specific CMakeList?

Automake is very difficult to make work for Windows, and often rewriting the buildsystem is the far easier path :)

>Is it worth submitting my changes as a PR?

_Absolutely_! We firmly believe in incremental progress -- as long as we avoid regressions, partial support is better than no support and someone else (or you ;)) can contribute better support in the future.

**Comment by mcraveiro** [comment]
<use automake>
> Automake is very difficult to make work for Windows, and often rewriting the buildsystem is the far easier path :)

OK understood :-) thanks for the feedback @ras0219-msft 

<submit PR>
> _Absolutely_! We firmly believe in incremental progress -- as long as we avoid regressions, partial support is better than no support and someone else (or you ;)) can contribute better support in the future.

Excellent. However, ran into additional problems; I just decided to do a build on a clean machine using Ubuntu Trusty (the image used in Travis), it now appears ODB cannot find its headers (config in particular). Let me get to the bottom of this first and then submit a PR.

Cheers

**Comment by mcraveiro** [comment]
Hm, not sure why but the formatting of the reply is coming out all weird!

**Comment by Praying** [comment]
> > Is it not possible to use the original automake files instead of creating a vcpkg specific CMakeList?
> 
> Automake is very difficult to make work for Windows, and often rewriting the buildsystem is the far easier path :)
> 
> > Is it worth submitting my changes as a PR?
> 
> _Absolutely_! We firmly believe in incremental progress -- as long as we avoid regressions, partial support is better than no support and someone else (or you ;)) can contribute better support in the future.

Hello ,now the libodb and libodb-mysql in vcpkg is 2.4 version, and when it comes to c++17 will not compile successful. Is it possible to upgrade it to 2.5.0-b9 which is the latest version.

**Comment by mcraveiro** [comment]
Drats, I am trying to upgrade all my stuff to C++ 17 at the moment so I'll be hit with this problem as well :-( I'll give it a go once the main problems have been addressed. I take it you are on Windows @Praying?

**Comment by Praying** [comment]
> Yes, I'm on the same boat as you. I think our best bet may just be to bite the bullet and try update vcpkg to build 2.5.0-b9 - let's see what breaks when we do that :-) I'll give it a spin and report back. However, I'm on the process of updating all of my build infrastructure to use vcpkg (Linux, OSX and Windows) instead of a homegrown/manual zips hackery that I had, so it may take me a few days to get to it. If you make any progress in the mean time do let me know.

Any way can we chat? emails? or Instant message? Telegram? I am in Chinese, and don't know what software that popular aboard.

**Comment by mcraveiro** [comment]
You are welcome to come up to the gitter of my project [1], though I must say I tend to reply only when I have some downtime between builds etc :-)

[1] https://gitter.im/MASD-Project/Lobby

**Comment by Praying** [comment]
@mcraveiro 
[functions.zip](https://github.com/Microsoft/vcpkg/files/2461534/functions.zip)

**Comment by mcraveiro** [comment]
Thanks @Praying as discussed over on gitter, I will get to it over the next couple of days and see if I can integrate your findings. Cheers

**Comment by mcraveiro** [comment]
I still haven't got round to working on this fix - build is broken in far more interesting ways, sadly :-D but if anyone is interested, I've moved the admittedly trivial patch over to its own branch [1] so that I can keep rebasing vcpkg. However, I have not yet looked into @Praying's work, nor will I have time until next week by the look of things.

[1] https://github.com/MASD-Project/vcpkg/tree/libodb

**Comment by Praying** [comment]
@mcraveiro  Try what I have done.
https://github.com/Praying/ODB2.5.0_Vcpkg_Config

**Comment by mcraveiro** [comment]
Right, for anyone interested in this, I *finally* had the time to sort out my issues with ODB on Linux. Well, at least it compiles and installs. It also adds support for ```libodb-boost```. However, I haven't cleaned up my changes - for now, I'm using a "generic" branch where I work simultaneously on all issues affecting all packages I'm interested in. Its really not ideal to raise a PR but it may be of use if you need ODB right now. The branch is available here [1]. 

Disclaimer: I'm not entirely sure when I'll get a chance to create a clean branch suitable for a PR, nor do I think this is the "correct" approach - I literally just copied the ```config.h``` files used in the Debian Testing packages into the ports and reused them via ```configure_file```. I suspect the ODB devs will not fall in love with this approach :-) However, it seems other ports are doing similar things (I grepped for ```configure_file``` to get a feel) and, If nothing else, it has the advantage of behind an improvement on current master which is in line with @ras0219-msft's suggestions above [2].

Finally, I did not have yet the time to review @Praying's changes [3], as they are more focused on bringing ODB to the beta release 2.5.0-b9. I haven't yet encountered problems with 2.4, so I'll be sticking with that for now until it breaks.

[1] https://github.com/DomainDrivenConsulting/vcpkg/commits/ddc
[2] https://github.com/Microsoft/vcpkg/issues/4428#issuecomment-428164743
[3] https://github.com/Praying/vcpkg/tree/update_odb_to_2.5.0-b9

**Comment by mcraveiro** [comment]
OK I finally remembered why I wanted ODB 2.5 - its needed for C++ 17 support [1]. Due to this I decided to stop being lazy and created a clean branch for the 2.4 changes with associated PR [2]. I won't PR the ODB 2.5 changes until ODB is released I don't think.

Also, note that at present ```libodb-mysql``` does not compile because the MySQL library it depends on does not compile. This is discussed in this issue [3].

I will close the ticket once the PR [2] is merged.

[1] https://www.codesynthesis.com/pipermail/odb-users/2018-March/003972.html
[2] https://github.com/Microsoft/vcpkg/pull/5260
[3] https://github.com/Microsoft/vcpkg/issues/5046

**Comment by mcraveiro** [comment]
For anyone who is interested in ODB 2.5: I have updated my branch [1] to this version, based on the original work from @Praying [2], [3]. However, after a closer inspection, I just noticed that @Praying's repo was all the way up to version ```b9``` whereas mine is only ```b3```, which is the latest available version in Boris personal package folder [4]. The code synthesis git repo seems to be at ```v2.5.0-b.9```. I will try to see if there is some quick hack available in vcpkg that allows me to download a tag from git rather than a tarball.

[1] https://github.com/DomainDrivenConsulting/vcpkg/tree/ddc
[2] https://github.com/Praying/ODB2.5.0_Vcpkg_Config/blob/master/libodb/portfile.cmake
[3] https://github.com/Praying/vcpkg/tree/update_odb_to_2.5.0-b9
[4] https://www.codesynthesis.com/~boris/tmp/odb/pre-release/
[5] https://git.codesynthesis.com/cgit/odb/libodb-pgsql/

**Comment by mcraveiro** [comment]
I have now updated my branch [1] to ODB 2.5.0.b9. I had a good look at @Praying's work, but my approach ended up differing a fair bit from his. Summary:

- used ```vcpkg_from_git``` to obtain the ODB tag ```v2.5.0-b.9``` directly from ODB's git repository.
- c++ 17 build: ```set(CMAKE_CXX_STANDARD 17)```
- GLIBC ABI: ```add_definitions(-D_GLIBCXX_USE_CXX11_ABI=1)```
- copied manually the generated files for ```config.h``` (all) and ```options.{chi}xx``` (pgsql and sqlite). I copied these files from the 2.4.0 packages - which I think its OK because I did not see any significant changes to the input files ```options.cli``` (only a copyright update).

Libraries compile and install cleanly on Linux x64. I consider my work here done so I'll close the ticket.

Thanks

[1] https://github.com/DomainDrivenConsulting/vcpkg/tree/ddc
